File: Cpp/Cheapest Flights Within K Stops.cpp
// 20ms, 79.97%; 11MB, 100.00%
static int x = []() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    return 0;
}();

class Solution
{
public:
    int findCheapestPrice(int n, const vector<vector<int>> &flights, int src, int dst, int K)
    {
        vector<int> price(n, 2e4);
        price[src] = 0;
        vector<int> tmp(price);
        for (int i = 0; i <= K; ++i) {
            for (const auto &f : flights) tmp[f[1]] = min(tmp[f[1]], price[f[0]] + f[2]);
            price = tmp;
        }
        return (price[dst] == 2e4 ? -1 : price[dst]);
    }
};

File: Python3/Cheapest Flights Within K Stops.py
# 112ms, 28.56%; 13.6MB, 49.54%
class Solution:
    def findCheapestPrice(self, n: int, flights: list, src: int, dst: int, K: int) -> int:
        price = [2e4]*n
        price[src] = 0
        for _ in range(K + 1):
            tmp = price[:]
            for u, v, w in flights:
                tmp[v] = min(tmp[v], price[u] + w)
            price = tmp
        return price[dst] if price[dst] < 2e4 else - 1
